tests/checkout/typechange.c

```diff
@@ -289,6 +290,10 @@ void test_checkout_typechange__checkout_with_conflicts(void)
 		cl_assert_equal_i(cts.updates, 0);
 		cl_assert_equal_i(cts.ignored, 0);
 
+		cl_git_pass(git_repository_index(&index, g_repo));
+		cl_git_pass(git_index_read(index, 1));
```

Comment:
These interspersed changes of `git_index_read` leave me a bit uneasy. Does it mean that users of libgit2 may now face  places where they have to add `git_index_read` or face errors otherwise? If so, we at least have to provide an update guide detailing in which exact situations this is now necessary, and even then people will complain.

Reply:
This is a good question.  This affects what happens when you are replacing the contents of the index (eg, checkout, merge) and you have unsaved changes, then you will need to save them (`git_index_write`) or revert them (`git_index_read`).  If you do not, this new code path will raise an error.

I don't think that this is a common scenario (except in our tests, since they play fast and loose with things).  But there is a very valid case that we would be breaking:  checking out with the `FORCE` flag or a reset with the `HARD` flag when you have unsaved changes in the index.  At the moment, you can do:

    git_index_add(...);
    /* no save */
    git_reset(..., GIT_RESET_HARD);

And your unsaved changes are ignored - the index is updated to match the checkout / reset target.  With this patch, this will be an error.

Increasingly, I think this is incorrect and that we should allow `git_checkout` / `git_reset` to operate when forcing.  I think that this would fix most of the _intentional_ use cases of this sort of behavior for people who were in a bad place but `reset` their way out of it, knowingly.

I'd also be happy to add a configurable option here for people who need an escape hatch to opt-in to the prior behavior.  It's also possible that this is too heavy handed and we should keep the default as-is.

I'll follow up a commit that updates the `reset` / `checkout` behavior to allow forcing in this instance.